src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/StringBuilderAppendShouldNotTakeSubstring.cs

```diff
@@ -0,0 +1,122 @@
+            helpLinkUri: "", // TODO (might require an addition to docs repository?
+            customTags: WellKnownDiagnosticTags.AnalyzerException);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(RuleReplaceOneParameter, RuleReplaceTwoParameter);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterOperationAction(this.Action, OperationKind.Invocation);
+        }
+
+        private void Action(OperationAnalysisContext context)
+        {
+            if (context.Operation is IInvocationOperation invocation)
+            {
+                var invokedMethod = invocation.TargetMethod;
+
+                if (invokedMethod.ContainingSymbol is INamedTypeSymbol namedType
+                    && namedType.Name == nameof(StringBuilder) 
+                    && namedType.ContainingNamespace.Name == nameof(System.Text)
+                    && namedType.ContainingNamespace.ContainingNamespace.Name == nameof(System)
+                    && namedType.ContainingNamespace.ContainingNamespace.ContainingNamespace.IsGlobalNamespace
+                    && invokedMethod.Name == nameof(StringBuilder.Append))
+                {
+                    var parameters = invokedMethod.Parameters;
+                    if (parameters.Length == 1 && parameters[0].Type.SpecialType == SpecialType.System_String)
+                    {
+                        IArgumentOperation argument = invocation.Arguments.FirstOrDefault();
+                        if (argument.Value is IInvocationOperation invocationExpression
+                            && invocationExpression.TargetMethod is IMethodSymbol parameterMethod
+                            && parameterMethod.Name == nameof(String.Substring)
+                            && parameterMethod.ContainingSymbol is INamedTypeSymbol parameterMethodOwner
+                            && parameterMethodOwner.SpecialType == SpecialType.System_String)
+                        {
+                            switch (parameterMethod.Parameters.Length)
+                            {
+                                case 1:
+                                    context.ReportDiagnostic(
+                                        Diagnostic.Create(
+                                            RuleReplaceOneParameter, 
+                                            invocation.Syntax.GetLocation()));
+                                    break;
+                                case 2:
+                                    context.ReportDiagnostic(
+                                        Diagnostic.Create(
+                                            RuleReplaceTwoParameter,
+                                            invocation.Syntax.GetLocation()));
```

Comment:
i don't really understand the need for hte different diagnostic IDs here.

Reply:
when do I need different diagnostic IDs - and when I don't?
Currently the fixer depends on the diagnostics-ID to apply one or the other fix. That could be done inside the code of course instead...

Reply:
> when do I need different diagnostic IDs - and when I don't?

Usually if you need to give the user a different message.  But, IMO, these messages should fairly short. Like "Avoid unnecessary string allocations when building strings".  You don't need to hyper-specialize the messages for these small differences.

Reply:
+1 This seems like something that can be changed to a single diagnostic. However, feel free to keep your implementation as-is and add a note to the tracking bug for TODO items after your PR and we can clean it up. If you are up to cleaning it up in this PR itself, then go for it.

> Currently the fixer depends on the diagnostics-ID to apply one or the other fix. That could be done inside the code of course instead

Yep, that is preferable instead of generating different diagnostic IDs for end users.

---
In reply to: [225003422](https://github.com/dotnet/roslyn-analyzers/pull/1854#discussion_r225003422) [](ancestors = 225003422)

Reply:
added a note to the issue https://github.com/dotnet/roslyn-analyzers/issues/1897